app/services/otp_service.rb

```diff
@@ -0,0 +1,60 @@
+class OtpService
+  # This service largely duplicates functions of the two_factor_authentication gem
+  # However, as it makes assumptions as to the field to validate for direct otps,
+  # this is required to extend that functionality to other fields.
+
+  cattr_accessor :direct_otp_length
+  cattr_accessor :direct_otp_valid_for
+
+  @@direct_otp_valid_for = Rails.configuration.devise.direct_otp_valid_for
+  @@direct_otp_length = Rails.configuration.devise.direct_otp_length
+
+
+  def initialize(user)
+    @user = user
+  end
+
+
+  def create_otp(field)
+    create_direct_otp(field)
+  end
+
+
+  def authenticate_otp(code_field, code)
+    direct_otp = @user.send(code_field)
```

Comment:
May be worth adding a comment to this method "Never under any circumstances call with a user provided code_field argument" or something to that effect.